### PR TITLE
[wip]expose get url method for wrapped signal

### LINKED ISSUE #22: expose method to get url from signalName + parameters
We should be able to make traditional links based on routing config since it allows to use `rightClick + copy` or `rightClick + open in new tab` actions.

It could be achieved with method to get url from signalName and parameters. So it would be able to create something like `Link` component @jeroenverfallie mentioned in #14 and without extra router configuration
```jsx
<Link signal="someSignal" args={ { someParam: true } } >
```
 which will render to 
```jsx
<a href="/#/url/param" onClick=(e) = { e.preventDefault(); this.props.signals.someSignal( { someParam: true })}>
```

### PR DESCRIPTION
**Pull request by Guria**
resolves #22 
please review

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -23,6 +23,37 @@ function router (controller, routes, options) {
...
+          return input;
+        }, input);
+      }
+
+      // Create url based on direct signal input or
+      // params passed from addressbar
+      url = params.reduce(function (url, param) {
+        var key = param.substr(1, param.length);
+        if (!(key in input)) {
+          throw new Error('Cerebral router - The signal "' + routes[route] + '" is bound to "' + route + '" route, but required param "' + key + '" wasn\'t provided.');
```
Comment (Guria) [inline index.js]: should we throw the same when `controller.signals.test.getUrl()` called without required params?

Comment (christianalfoni) [inline index.js]: Hm, you could do a try/catch around at call site of exposed getUrl below. Catch the error and produce a more specific url to the use of that method.

Comment (Guria) [inline index.js]: What you mean by `more specific url`?

Comment (christianalfoni) [inline index.js]: haha, error, not url... sorry :-)

Comment (Guria) [inline index.js]: Can't find a good way to do it. If I catch an error in exposed `getUrl` I am loosing information about which key wasn't provided. I think a message is generic enough to be thrown in both cases.

Comment (christianalfoni) [inline index.js]: Ah, yeah, good point. Though I would not say : "is falsy", I would say "property is defined", as the value can be false, null... even undefined if you would set that explicitly. We are checking for the existence of the key

Comment (Guria) [inline index.js]: This message have nothing about `is falsy`

Comment (Guria) [inline index.js]: https://github.com/christianalfoni/cerebral-router/pull/32/files#diff-168726dbe96b3ce427e7fedce31bb0bcR108 this message does.
But if key would falsy resulted url wouldn't match route. So error suggest to check provided value for key.

Comment (christianalfoni) [inline index.js]: Ah, I was looking at the one below...  sorry, 2 sec

File: index.js [hunk 2]
```diff
@@ -59,47 +91,23 @@ function router (controller, routes, options) {
...
-        urlMapper(url, checkRoute);
+      urlMapper(url, checkRoute);
 
-        if (!urlMatched) {
-          throw new Error('Cerebral router - Computed url for signal "' + routes[route] +'" can\'t match given route "' + route + '".\n' +
-                          'Check required params provided to signal is not falsy.');
-        }
+      if (!urlMatched) {
+        throw new Error('Cerebral router - Computed url for signal "' + routes[route] +'" can\'t match given route "' + route + '".\n' +
+                        'Check required params provided to signal is not falsy.');
```
Comment (christianalfoni) [inline index.js]: Here, should state something about it being missing key too

Comment (Guria) [inline index.js]: `Check required params was provided to signal and none of them is falsy.` ?

**Comment by Guria** [comment]
Let assume we have a route `/:param` for test signal. And someone call `controller.signals.test.getUrl({ param: 1, foo: bar })`.
Should we warn that foo parameter doesn't affect resulted url?

File: index.js (hunk 2 above)
Comment (christianalfoni) [inline index.js]: Yes, should be something like the above? "... there are missing keys in the params or no params passed". As:

```js
{
  show: false
}
```
Is a valid params to url: /popup/show/false (/popup/show/:state). Bad example, but yeah, it is supported :-)

Comment (Guria) [inline index.js]: Hmm. Shouldn't it provided as string? But your example is more dev friendly. Would check it and add to test cases.

Comment (Guria) [inline index.js]: @christianalfoni seems you have missed that I have changed it in #23 
Falsy values are replaced with empty string.
Use case you have provided seems useful while passing params to signal. I can white list to make `false` appear in url. But again, I found a bit confusing that restored param from url would be a string`'false'` which is `true` in fact. And developer still have to check `input.show !== 'false' || input.show`.

Now my strong suggestion is:
* allow ONLY string values for params used in route definition
* make Boolean values work but log out `console.warn` deprecations

**Comment by Guria** [comment]
pr doesn't ready closing for now

**Pull request closed without merging**